File: src/main/java/com/footverse/cart/dto/CartResponse.java
package com.footverse.cart.dto;

import java.math.BigDecimal;
import java.util.List;

/**
 * The caller's cart with its server-computed aggregates (dto-spec §12). The client never sends or
 * recomputes {@code subtotal} or {@code itemCount}: the server is the single source of truth for
 * money (dto-spec §1).
 *
 * @param items     the cart lines; empty when the caller has no cart yet, or an empty one
 * @param subtotal  the sum of every line total, computed by the server
 * @param itemCount the sum of every line's quantity (the cart badge value), <em>not</em> the number
 *                  of distinct lines (business-rules → Shopping Cart)
 */
public record CartResponse(
        List<CartItemResponse> items,
        BigDecimal subtotal,
        int itemCount) {
}

File: app/lib/features/cart/models/cart_response.dart
import 'package:json_annotation/json_annotation.dart';

import '../../product/models/decimal_json.dart';
import 'cart_item_response.dart';

part 'cart_response.g.dart';

/// The caller's cart with server-computed aggregates (dto-spec §12).
///
/// [subtotal] and [itemCount] are computed by the server and rendered exactly
/// as delivered; the client never recomputes money or aggregates (dto-spec §1;
/// business-rules). [itemCount] is Σ quantity of all lines — the cart badge
/// value — not the distinct-line count (`items.length`).
@JsonSerializable(createToJson: false)
class CartResponse {
  const CartResponse({
    required this.items,
    required this.subtotal,
    required this.itemCount,
  });

  factory CartResponse.fromJson(Map<String, dynamic> json) =>
      _$CartResponseFromJson(json);

  final List<CartItemResponse> items;
  @JsonKey(fromJson: decimalFromJson)
  final double subtotal;
  final int itemCount;
}
